socialstar.rb

```diff
@@ -0,0 +1,263 @@
+                                            message: "Gender must be one of #{Socialstar::GENDERS.join(' ')}" },
+                                            allow_nil: true
+  validates :payment_method,   inclusion: { in: Socialstar::PAYMENT_METHODS,
+                                            message: "Payment method must be one of #{Socialstar::PAYMENT_METHODS.join(' ')}" },
+                                            allow_nil: true
+  validates :fb_user_id, :uniqueness => { :message => "Facebook account is already registered", :allow_nil => true }
+
+  validate :valid_blog_categories
+  validate :parse_date_of_birth
+  validate :older_than_13_years
+
+  before_validation :upcase_address_state
+  before_save :set_country_code
+  before_create :convert_follower_counts
+  after_create :assign_invitation_code
+  after_update :track_status_change, :if => :status_changed?
+
+  def set_country_code
+    if CA_STATES.include?(address_state)
+      self.country_code = "CA"
+    else
+      self.country_code = "US"
+    end
+  end
+
+  def update_attributes_from_auth_hash!(auth_hash)
+    self.fb_user_id                             = auth_hash.uid
+    self.provider                               = auth_hash.provider
+    self.social_media_access_tokens['facebook'] = auth_hash.credentials.to_hash.symbolize_keys
+    self.password                               = Devise.friendly_token[0, 20]
+    self.status                                 = :accepted
+    save!
+  end
+
+  def assign_invitation_code!(code)
+    self.invitation_code = InvitationCode.where(:code => code).first
+    self.save!
+  end
+
+  def attributes_for_crowdtap
+    attributes = { :application_id => id }
+    CROWDTAP_ATTRIBUTES.each do |attribute|
+      attributes[attribute] = send(attribute)
+    end
+    attributes
+  end
+
+  def age
+    (Time.now.to_date - self.date_of_birth).to_i/365 if self.date_of_birth
+  end
+
+  def visit
+    tracker.set!
+  end
+
+  private
+
+  def older_than_13_years
+    if date_of_birth && date_of_birth > 13.years.ago.to_date
+      errors.add(:date_of_birth, 'You must be 13 years or older to sign up')
+    end
+  end
+
+  def parse_date_of_birth
+    begin
+      dob = attributes_before_type_cast["date_of_birth"]
+      self.date_of_birth = DateTime.strptime("#{dob}", "%m/%d/%Y") if dob && dob.is_a?(String)
+    rescue ArgumentError => e
+      if e.message =~ /invalid date/
+        errors.add(:date_of_birth, "Not a valid date")
+      end
+    end
+  end
+
+  def valid_blog_categories
+    blog = social_channels["blog"]
+    if blog && blog["categories"]
+      blog["categories"].any? do |category|
+        unless Socialstar::BLOG_CATEGORIES.include?(category)
+          errors.add(:social_channels, "Blog categories must be one of #{Socialstar::BLOG_CATEGORIES.join(' ')}")
+        end
+      end
+    end
+  end
+
+  def convert_follower_counts
+    social_channels.each do |channel, fields|
```

Comment:
Unused block argument - `channel`. If it's necessary, use `_` or `_channel` as an argument name to indicate that it won't be used.